core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java

```diff
@@ -982,11 +982,25 @@ void propagateCostImprovements(RelNode rel) {
         if (!relNode.getTraitSet().satisfies(subset.getTraitSet())) {
           continue;
         }
-        if (!cost.isLt(subset.bestCost)) {
+
+        // Update subset best and best's cost when we find a cheaper rel
+        if (relNode != subset.best && !cost.isLt(subset.bestCost)) {
           continue;
         }
-        // Update subset best cost when we find a cheaper rel or the current
-        // best's cost is changed
+
+        // Update the RelNode's cost when we find that the cost is changed.
+        //
+        // Why do we need to update it?
+        // When RelSet's one of the subsets find a cheaper RelNode, we need to update the
+        // parents of the subset to have the best RelNode and best cost.
+        // In theory, this cost will become smaller, But according to
+        // The SQL we added in the JdbcAdapterTest {@link testVolcanoPlannerInternalValid},
+        // it shows RelNode's cost will become bigger sometimes.
+        // So we update it.
```

Comment:
I'm not sure if the keyword "we" is appropriate to describe

Reply:
Yes, I think comments should be impersonal - if you want to say "I" you should probably sign it in the code as well.

The comment could be just `cost X can be bigger than cost Y, as shown by the test @link ...`.

Reply:
@mihaibudiu  I think the java doc is ok now, WDYT?